packages/block-editor/src/store/reducer.js

```diff
@@ -0,0 +1,895 @@
+	if ( original === working ) {
+		return { ...original };
+	}
+
+	return working;
+}
+
+/**
+ * Returns true if the two object arguments have the same keys, or false
+ * otherwise.
+ *
+ * @param {Object} a First object.
+ * @param {Object} b Second object.
+ *
+ * @return {boolean} Whether the two objects have the same keys.
+ */
+export function hasSameKeys( a, b ) {
+	return isEqual( keys( a ), keys( b ) );
+}
+
+/**
+ * Returns true if, given the currently dispatching action and the previously
+ * dispatched action, the two actions are updating the same block attribute, or
+ * false otherwise.
+ *
+ * @param {Object} action         Currently dispatching action.
+ * @param {Object} previousAction Previously dispatched action.
+ *
+ * @return {boolean} Whether actions are updating the same block attribute.
+ */
+export function isUpdatingSameBlockAttribute( action, previousAction ) {
+	return (
+		action.type === 'UPDATE_BLOCK_ATTRIBUTES' &&
+		action.clientId === previousAction.clientId &&
+		hasSameKeys( action.attributes, previousAction.attributes )
+	);
+}
+
+/**
+ * Returns true if, given the currently dispatching action and the previously
+ * dispatched action, the two actions are modifying the same property such that
+ * undo history should be batched.
+ *
+ * @param {Object} action         Currently dispatching action.
+ * @param {Object} previousAction Previously dispatched action.
+ *
+ * @return {boolean} Whether to overwrite present state.
+ */
+export function shouldOverwriteState( action, previousAction ) {
+	if ( ! previousAction || action.type !== previousAction.type ) {
+		return false;
+	}
+
+	return isUpdatingSameBlockAttribute( action, previousAction );
+}
+
+/**
+ * Higher-order reducer targeting the combined editor reducer, augmenting
+ * block client IDs in remove action to include cascade of inner blocks.
+ *
+ * @param {Function} reducer Original reducer function.
+ *
+ * @return {Function} Enhanced reducer function.
+ */
+const withInnerBlocksRemoveCascade = ( reducer ) => ( state, action ) => {
+	if ( state && action.type === 'REMOVE_BLOCKS' ) {
+		const clientIds = [ ...action.clientIds ];
+
+		// For each removed client ID, include its inner blocks to remove,
+		// recursing into those so long as inner blocks exist.
+		for ( let i = 0; i < clientIds.length; i++ ) {
+			clientIds.push( ...state.blocks.order[ clientIds[ i ] ] );
+		}
+
+		action = { ...action, clientIds };
+	}
+
+	return reducer( state, action );
+};
+
+/**
+ * Higher-order reducer which targets the combined blocks reducer and handles
+ * the `RESET_BLOCKS` action. When dispatched, this action will replace all
+ * blocks that exist in the post, leaving blocks that exist only in state (e.g.
+ * reusable blocks) alone.
+ *
+ * @param {Function} reducer Original reducer function.
+ *
+ * @return {Function} Enhanced reducer function.
+ */
+const withBlockReset = ( reducer ) => ( state, action ) => {
+	if (
+		state &&
```

Comment:
The fact we check for a truthy state implies (incorrectly) that we'd handle the case where it's falsey from within the reducer. But the new block editor reducer does not have a case for `RESET_BLOCKS` nor `INIT_BLOCKS`, thus the following would not work at all as expected:

```js
editor( undefined, {
	type: 'RESET_BLOCKS',
	blocks: [ { clientId: 'bananas', innerBlocks: [] } ],
} );
```